tests/user_input_checks_test.py

```diff
@@ -435,3 +436,43 @@ def test_passing_custom_density_estimator(arg):
         density_estimator = arg
     prior = MultivariateNormal(torch.zeros(2), torch.eye(2))
     _ = SNPE_C(prior=prior, density_estimator=density_estimator)
+
+
+def test_validate_theta_and_x_cpu():
+
+    cpu_device = torch.device('cpu')
+
+    theta = torch.ones((32,8), dtype=torch.float32).to(cpu_device)
+    x = torch.zeros((32,100), dtype=torch.float32).to(cpu_device)
+    otheta = torch.FloatTensor((32,8)) #using an explicit type
+
+    #check assumptions in implementation
+    assert isinstance(theta, torch.Tensor)
+    assert theta.dtype == torch.float32
+    assert isinstance(theta, torch.FloatTensor)
+    assert otheta.dtype == torch.float32
+
+    # test on cpu
+    validate_theta_and_x(theta, x)
+
+    with pytest.raises(AssertionError) as exc:
+        validate_theta_and_x(theta, x.to(torch.float64))
+
+
+@pytest.mark.gpu
```

Comment:
I know this goes in the other direction of what @michaeldeistler suggested, sorry, but with this split we have an almost 100% code repetition. 
Could we just ignore the mark.gpu here and use a single test with a device parameter and `mark.parametrize("device", ("gpu", "cpu"))` and then use the "process_device" function or an if-else like Peter did in line 465?

Reply:
the if-else would return `"cpu"` if `torch.cuda.is_available()` is `False` and then just run twice. 

Or it would just "pass" when `torch.cuda.is_available()` is `False`.

Reply:
@janfb thanks for the feedback. I agree that code dublication is a bad thing. Moreover, the two versions of the code look almost identical except for one `assert` that is crucial to illustrate the behavior. This is why, I had a `pytest.skip` clause added just before the gpu relevant code snippets.

Please let me know how to move forward, I'd love to merge this PR as it is holding me back. I am forced to use cpu-based training on my dataset at this moment which appears to be somewhat slower. I'd be curious to compare it to gpu based training.

Reply:
OK, I suggest the following (how can I suggest a change like you did in that other PR?): 
```python
from sbi.utils.torchutils import process_device
@pytest.mark.parametrize("device", ("cpu", "cuda:0"))
def test_validate_theta_and_x(device):

     # Fallback to "cpu" if "gpu" not available.
     device = process_device(device)

     theta = torch.ones((32,8), dtype=torch.float32).to(device)
     x = torch.zeros((32,100), dtype=torch.float32).to(device)
     otheta = torch.FloatTensor((32,8)) #using an explicit type

     #check assumptions in implementation
     assert isinstance(theta, torch.Tensor)
     assert theta.dtype == torch.float32
     assert isinstance(theta, torch.FloatTensor)
     if device == "cpu":
         assert otheta.dtype == torch.float32

     validate_theta_and_x(theta, x)

     with pytest.raises(AssertionError) as exc:
         validate_theta_and_x(theta, x.to(torch.float64))
```

Reply:
or add 
```python
if decive=="cuda:0" and not torch.cuda.is_available(): 
    pass
```
to avoid running the test twice.

Reply:
I can do that, but this would not illustrate the core point of the test and the root cause of the bug I suggest to fix. This line:
```
     assert isinstance(theta, torch.FloatTensor)
```
only works on the CPU. the assertion fails on the gpu!

Reply:
Yes, of course. Sorry, I just missed the `not` after the `assert`. 
OK, then I am fine with just keeping both versions of the test, or you make the `assert` work with both scenarios by adding another if-else on device. How you prefer.

Reply:
Err, I am a bit lost now. From my point of view, I'd love the fix to get into `main` as soon as possible. This is my highest priority, so I can move on.

At the end of the day, the current version of github actions workflow relies on the feature @michaeldeistler suggested earlier, so that you can use `pytest -m "not slow and not gpu" tests/ --cov=sbi --cov-report=xml` there. If the `@pytest.mark.parametrize("device", ("cpu", "cuda:0"))` does support that, fine with me. Do you know @janfb ?

Reply:
yes, the only difference is that the `pytest.mark.gpu` would actually skip the test during the github workflow, while my suggestion would just bypass it by setting device to "cpu" or by letting the "gpu" test just "pass". 
So should be fine with just using  `@pytest.mark.parametrize("device", ("cpu", "cuda:0"))` and taking care that the test works locally with gpu.